### PR TITLE
feat(workflow): add actions cost audit guidance

### PR DESCRIPTION
**Pull request by lhpaul**
## Summary

Implements #1099 by adding a lightweight GitHub Actions cost-audit helper and the accompanying guidance for retrospectives and downstream template-sync reviews.

What changed:

- Added `scripts/development-workflow/actions-cost-audit.sh` to aggregate recent workflow run count and wall time by workflow using `gh run list`.
- Added `scripts/development-workflow/tests/test-actions-cost-audit.sh` with mocked `gh` coverage for aggregation, incomplete data, empty data, permission failures, recommendation outcomes, public/private framing, and input validation.
- Added `docs/workflow/development-workflow/integrations/actions-cost-audit.md` and linked it from the workflow README, e2e regression guide, PR-Agent guide, and scripts README.
- Added the #1099 `[Unreleased]` CHANGELOG entry.

Spec: `docs/specs/developments/20260701091532_1099-actions-cost-audit-guidance/1_1099-actions-cost-audit-guidance_specs.md`
Plan: `docs/specs/developments/20260701091532_1099-actions-cost-audit-guidance/2_1099-actions-cost-audit-guidance_implementation-plan.md`
Smoke runbook: `docs/testing/workflow/1099-actions-cost-audit-guidance.smoke-test.md`

## Verification

- `bash scripts/development-workflow/tests/test-actions-cost-audit.sh`
- `shellcheck --severity=warning scripts/development-workflow/actions-cost-audit.sh scripts/development-workflow/tests/test-actions-cost-audit.sh`
- `python3 scripts/lint/workflow-shell-guard-lint.py --base-ref origin/develop`
- `npx markdownlint-cli2 "scripts/development-workflow/README.md" "docs/workflow/development-workflow/README.md" "docs/workflow/development-workflow/integrations/actions-cost-audit.md" "docs/workflow/development-workflow/integrations/e2e-regression.md" "docs/workflow/development-workflow/integrations/pr-agent.md" "docs/testing/workflow/1099-actions-cost-audit-guidance.smoke-test.md" "CHANGELOG.md"`
- `find docs/specs/developments docs/testing/workflow -name "*.md" -print0 | xargs -0 python3 scripts/lint/markdown-heuristic-lint.py CHANGELOG.md`
- `bash scripts/lint/check-changelog-duplicate-headers.sh CHANGELOG.md`
- `git diff --check`
- `./scripts/development-workflow/actions-cost-audit.sh --limit 10`

## Pre-Submission Self-Review

Stale markers: none found in added staged lines.
Caller/sibling consistency: verified helper name, doc links, README index entry, smoke runbook expectations, and CHANGELOG entry match.
Coverage: all #1099 acceptance criteria are addressed by the helper output, integration guide, mocked tests, and smoke runbook.
Plan deviations: none.

## Test Harness Coverage Checklist

- Empty / zero-length input: checked via mocked empty `gh run list` response.
- Whitespace-only input: checked via whitespace-only `--limit` validation.
- Boundary values: checked via `--limit 0` validation and normal positive limit runs.
- Missing / absent environment variables: not applicable; helper has no required environment variables.
- Concurrent / parallel execution: not applicable; helper is read-only and test fixtures are isolated in `mktemp -d`.
- Negative assertions: checked via permission failure, invalid limit, invalid timestamp, and since-filter exclusion assertions.
- Single EXIT trap: checked; harness uses one cleanup trap.
- Variable-length quoting safety: checked via `--repo "owner/repo with space"`.
- `BASH_SOURCE` / harness guard placement: not applicable; harness is executed directly and does not support sourced mode.
- Sourced-function ordering: not applicable; harness does not source other scripts.

## Script-Accuracy Self-Check

Script: `scripts/development-workflow/actions-cost-audit.sh`

- Options `--limit`, `--repo`, `--since`, `--format`: verified in source option parser and usage text.
- GitHub data source `gh run list`: verified in source command construction.
- JSON fields `workflowName`, run IDs, timestamps, status, conclusion, event, branch, URL: verified in source `FIELDS` list.
- Permission/auth failure path: verified in source `gh run list failed` error branch and focused test.
- Markdown output sections: verified in source render functions and focused test.
- Public/private framing and recommendation outcomes: verified in source render functions, docs, and focused test.

## CHANGELOG Preview

- **Actions cost-audit guidance** (#1099): Add lightweight workflow run-volume and wall-time audit guidance for retrospectives and downstream template-sync reviews.

Closes #1099

---

> [!NOTE]
> **Low Risk**
> Read-only shell helper and docs only; no workflow triggers, secrets, or billing mutations.
> 
> **Overview**
> Adds a **read-only maintainer helper** to review recent GitHub Actions **run volume and wall time** before changing workflow defaults, running retrospectives, or judging template-sync cost risk in private downstream repos.
> 
> `scripts/development-workflow/actions-cost-audit.sh` calls `gh run list` (no billing APIs), aggregates by workflow, and prints **markdown** with a summary table, data-limitation notes, public/private cost framing, and a recommendation worksheet (`keep`, `narrow`, `make opt-in`, etc.). Options include `--limit`, `--repo`, and `--since`.
> 
> Documentation and wiring: new `integrations/actions-cost-audit.md`, entries in the workflow and scripts READMEs, and pointers from the e2e-regression and PR-Agent integration guides. **`test-actions-cost-audit.sh`** mocks `gh` for aggregation, empty data, validation, and permission failures.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 823578dcb5a23c7463b533053b2f11982b2987c2. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

## Why Safe To Merge

```json
{
  "why_safe_to_merge": {
    "scope": "Adds one read-only GitHub Actions audit helper, one mocked shell test, workflow documentation links, and the #1099 CHANGELOG entry. It does not change workflow triggers, repository settings, labels, tracker state, or billing configuration.",
    "tests": "Focused mocked shell test, ShellCheck, workflow shell guard, markdownlint, markdown heuristic lint, changelog duplicate/header checks, git diff whitespace check, and live non-mutating audit smoke check all passed. The Bugbot duration finding was fixed and covered by a new all-incomplete workflow regression assertion.",
    "reviewer_outcome": "Internal code review approved. Automated reviewer loop reran after the Bugbot fix and returned clean with zero unresolved review threads.",
    "ci_outcome": "pr-ci-loop.sh for PR #1111 returned RESULT=green with no failing or pending checks after ready-for-regression was applied.",
    "rollback_or_cleanup_risk": "Low. Rollback is removing the helper, its focused test, docs links, integration guide, and CHANGELOG entry. The helper is read-only and has no persistent side effects."
  }
}
```

### CODE REVIEW

**Comment by lhpaul** [comment]
Internal code review gate: APPROVED

Findings: none.

Direct fixes applied during internal review: none.

Coverage reviewed:
- Spec compliance: all #1099 acceptance criteria are covered by the helper output, docs, mocked tests, and smoke runbook.
- Code quality: shell input validation, gh/jq failure handling, incomplete data handling, and non-mutating behavior are covered.
- Test harness checklist: covered in PR body and verified by focused tests.
- Script-accuracy checklist: documented claims match `scripts/development-workflow/actions-cost-audit.sh`.

Verdict: APPROVED

**Comment by lhpaul** [comment]
bugbot run

**Comment by lhpaul** [comment]
### Automated Reviewer Loop Summary

**Result:** clean — no blocking findings
**Platforms:** pr-agent (clean), bugbot (clean)
**Findings:** 0 blocking, 0 suggestions

**Ready reviewer phase:** No net-new blocker was found after the draft GitHub gate.
**Ready-phase platforms:** bugbot

**Advisory findings (non-blocking):**
- Possible Issue ([view comment](https://github.com/lhpaul/ai-dev-framework-template/pull/1111#issuecomment-4855578641))
- Possible Issue ([view comment](https://github.com/lhpaul/ai-dev-framework-template/pull/1111#issuecomment-4855578641))

*Posted automatically by `pr-review-loop.sh`.*

File: scripts/development-workflow/actions-cost-audit.sh [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+      "Runs included: " + ($filtered_count | tostring) + ". Incomplete duration records: " + ($incomplete_count | tostring) + ".\n\n" +
+      "## Workflow summary\n\n" +
+      "| Workflow | Runs | Completed | Incomplete duration | Total wall time | Avg wall time | Dominant events | Recent runs |\n" +
+      "| --- | ---: | ---: | ---: | ---: | ---: | --- | --- |\n" +
+      ($summary
+       | map("| " + (.workflow | safe_cell) +
+             " | " + (.runs | tostring) +
+             " | " + (.completed | tostring) +
+             " | " + (.incomplete | tostring) +
+             " | " + minutes(.totalSeconds) +
```
[the remark this answers is not in the document]
Comment (lhpaul) [reply scripts/development-workflow/actions-cost-audit.sh:215]: Fixed in commit 823578d. Workflow groups with no usable duration samples now render total and average wall time as `n/a`, and the focused test covers an all-incomplete workflow group.

**Comment by lhpaul** [comment]
bugbot run

**Comment by lhpaul** [comment]
## /run-epic PR Disposition

- Scope source: epic #1095
- Item: #1099 - Add lightweight Actions cost-audit guidance
- PR: #1111
- Reviewed head SHA: `823578dcb5a23c7463b533053b2f11982b2987c2`
- Reviewer result: clean
- Blocking findings: 0
- Advisory findings: 2
- Risk: medium
- Merge authority: delegated -- mayMerge=true, maxRisk=high
- Final decision: merged

### Risk Reasons

- low-risk docs or tests change: CHANGELOG.md
- low-risk docs or tests change: docs/workflow/development-workflow/README.md
- low-risk docs or tests change: docs/workflow/development-workflow/integrations/actions-cost-audit.md
- low-risk docs or tests change: docs/workflow/development-workflow/integrations/e2e-regression.md
- low-risk docs or tests change: docs/workflow/development-workflow/integrations/pr-agent.md
- low-risk docs or tests change: scripts/development-workflow/README.md
- workflow script or harness change: scripts/development-workflow/actions-cost-audit.sh

### Invocation Policy

- Original command: `-epic 1095`
- Copy-paste equivalent: `/run-epic --epic 1095 --base develop-actions-cost-reduction --may-start-backlog true --delegate-review --may-merge --max-risk high`
- Confirmation: human confirmed; maxRisk changed to high

| Field | Recommended | Selected | Effective |
| --- | --- | --- | --- |
| mayStartBacklog | true | true | true |
| delegateReview | true | true | true |
| mayMerge | true | true | true |
| maxRisk | medium | high | high |
| base | develop-actions-cost-reduction | develop-actions-cost-reduction | develop-actions-cost-reduction |

### Checkpoint Policy

- Field source: effective
- Pending applicable checkpoints: 0

| Item | Stage | Domain | State | Reason | Required action |
| --- | --- | --- | --- | --- | --- |
| #1099 | spec | product | satisfied | cost-audit guidance scope needed spec approval | approve product-facing cost-audit guidance behavior |
| #1099 | plan | technical | satisfied | implementation approach and validation need plan approval | approve plan before implementation |
| #1099 | implementation | technical | satisfied | cost-audit workflow guidance may affect shared workflow recommendations | approve implementation before delegated merge |

### Advisory Decisions

| Source | Category | Decision | Rationale |
| --- | --- | --- | --- |
| PR-Agent | Possible Issue | accepted as non-blocking false positive | The advisory flagged jq injection via --since, but the script passes SINCE into jq with --arg and validates it with fromdateiso8601 before use; no shell interpolation into the jq program occurs. |
| PR-Agent | Possible Issue duplicate | deduplicated | The reviewer loop listed the same PR-Agent possible-issue comment twice; the disposition above covers the underlying advisory. |

### Verification Evidence

- Labels: ready-for-human-review, ready-for-regression, Review effort 2/5
- CI result: green
- Reviewer summary present: true
- Unresolved thread count: 0
- PR merge state: MERGED
- Issue state: CLOSED
- Project status: Merged

### Protocol Deviations

None.
